### PR TITLE
feat(core): control oo callbacks from request

### PR DESCRIPTION
**Pull request by Reinis-FRP**
**Motivation**

The Optimistic Oracle currently uses a try-catch structure to perform callbacks on the requester contract. However, such a structure is potentially vulnerable to gas griefing.

**Summary**

Only perform callbacks if the requester has registered its readiness to receive them.

**Details**

By default callbacks are disabled, but while in `Requested` state the requester can enable desired callbacks with new `setCallbacks` method.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [x]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

**Comment by Reinis-FRP** [comment]
Note that I was not able yet to test this due to "Stack too deep" error when compiling.

File: packages/core/contracts/oracle/implementation/OptimisticOracle.sol [hunk 1]
```diff
@@ -351,8 +369,8 @@ contract OptimisticOracle is OptimisticOracleInterface, Testable, Lockable {
         // End the re-entrancy guard early to allow the caller to potentially take OO-related actions inside this callback.
         _startReentrantGuardDisabled();
         // Callback.
-        if (address(requester).isContract())
-            try OptimisticRequester(requester).priceProposed(identifier, timestamp, ancillaryData) {} catch {}
```
Comment (nicholaspai) [inline packages/core/contracts/oracle/implementation/OptimisticOracle.sol:355]: I rly like removing try-catches

**Review by nicholaspai — APPROVED** [review]
This solution makes sense!

**Review by chrismaree — COMMENTED** [review]
makes sence! would this not then mean that this PR: https://github.com/UMAprotocol/protocol/pull/3935#pullrequestreview-989997783 becomes redundant as you _could_ setCallbacks_ there to enable this?

File: packages/core/contracts/oracle/implementation/OptimisticOracle.sol [hunk 2]
```diff
@@ -293,8 +298,35 @@ contract OptimisticOracle is OptimisticOracleInterface, Testable, Lockable {
...
+    /**
+     * @notice Sets which callbacks should be enabled for the request.
+     * @param identifier price identifier to identify the existing request.
+     * @param timestamp timestamp to identify the existing request.
+     * @param ancillaryData ancillary data of the price being requested.
+     * @param callbackOnPriceProposed whether to enable the callback onPriceProposed.
+     * @param callbackOnPriceDisputed whether to enable the callback onPriceDisputed.
+     * @param callbackOnPriceSettled whether to enable the callback onPriceSettled.
+     */
+    function setCallbacks(
```
Comment (chrismaree) [inline packages/core/contracts/oracle/implementation/OptimisticOracle.sol:314]: ^

**Review by mrice32 — APPROVED** [review]

**Pull request merged**